net-mgmt/pfSense-pkg-arpwatch/files/usr/local/pkg/arpwatch.inc

```diff
@@ -0,0 +1,239 @@
+}
+
+function arpwatch_clear_database() {
+	mwexec("/usr/bin/find ".ARPWATCH_LOCAL_DIR." -type f -name 'arp_*.dat' -delete");
+	mwexec("/usr/bin/find ".ARPWATCH_LOCAL_DIR." -type f -name 'arp_*.dat-' -delete");
+}
+
+function arpwatch_parse_database() {
+	global $config;
+
+	$arpwatch_config = $config['installedpackages']['arpwatch']['config'][0];
+	$active_interfaces = $arpwatch_config['active_interfaces'];
+
+	$entries = [];
+
+	foreach (explode(',', $active_interfaces) as $active_interface) {
+		$ifname = convert_friendly_interface_to_real_interface_name($active_interface);
+		$arp_file = arpwatch_get_arp_file($ifname);
+
+		$database = file($arp_file);
+
+		foreach ($database as $line) {
+			list($mac, $ip, $timestamp, $hostname) = explode("\t", trim($line));
+
+			$pattern = '/(^|:)([a-f0-9])(:|$)/';
+
+			while (preg_match($pattern, $mac)) {
+				$mac = preg_replace('/(^|:)([a-f0-9])(:|$)/', '${1}0${2}${3}', $mac);
+			}
+
+			$oui = strtoupper(substr($mac, 0, 8));
+			$matches = [];
+
+			if (preg_match("/^$oui\s+(.*)$/m", file_get_contents(ARPWATCH_LOCAL_DIR.'/ethercodes.dat'), $matches)) {
+				$vendor = $matches[1];
+			}
+			else {
+				$vendor = 'unknown';
+			}
+
+			$entry = [
+				'ifname' => $ifname,
+				'ifdescr' => strtoupper($active_interface),
+				'mac' => $mac,
+				'vendor' => $vendor,
+				'ip' => $ip,
+				'hostname' => '',
+				'timestamp' => '',
+			];
+
+			if ($hostname) {
+				$entry['hostname'] = $hostname;
+			}
+
+			if ($timestamp) {
+				$entry['timestamp'] = strftime('%c', $timestamp);
+			}
+
+			$entries[] = $entry;
+		}
+	}
+
+	usort($entries, function($e1, $e2){
+		if ($e1['ifdescr'] == $e2['ifdescr']) {
+			return 0;
+		}
+
+		return ($e1['ifdescr'] < $e2['ifdescr']) ? -1 : 1;
+	});
+
+	return $entries;
+}
+
+function arpwatch_get_database_entries() {
+	// update arp files
+	mwexec("/usr/bin/killall -ALRM arpwatch || exit 0");
+
+	// wait a bit
+	usleep(0.2 * 10**6);
+
+	return arpwatch_parse_database();
+}
+
+function arpwatch_install_sendmail_proxy() {
+	$sendmail_proxy =<<< 'EOF'
+#!/usr/bin/env php
+<?php
+
+require_once("notices.inc");
+
+$fd = fopen('php://stdin','r');
+$message = stream_get_contents($fd);
+fclose($fd);
+
+if (false !== $message) {
+	$subject = [];
+	preg_match('/^Subject: (.*)$/m', $message, $subject);
+	$message = preg_replace('/^From: .*$/m', '', $message);
+	$message = preg_replace('/^To: .*$/m', '', $message);
+	$message = preg_replace('/^Subject: .*$/m', '', $message);
+	$message = preg_replace("/^(\n){4}/", '', $message);
+
+	send_smtp_message($message, "{$config['system']['hostname']}.{$config['system']['domain']} - Arpwatch Notification : {$subject[1]}");
+}
+
+?>
+
+EOF;
+
+	file_put_contents(ARPWATCH_SENDMAIL_PROXY, $sendmail_proxy);
+	chmod(ARPWATCH_SENDMAIL_PROXY, 0755);
+	
+	if (!file_exists(ARPWATCH_SENDMAIL_PATH)) {
+		@symlink(ARPWATCH_SENDMAIL_PROXY, ARPWATCH_SENDMAIL_PATH);
```

Comment:
Why is it necessary?

Reply:
@rbgarga  If it's anything like arpwatch was in the past, it's hardcoded to use `/usr/sbin/sendmail` and there is no way to change it.

Reply:
Exactly, I wasn't sure whether you would allow that but the path to sendmail is still hardcoded.